### PR TITLE
Enhancements to map_to_list processor

### PR DESCRIPTION
**Pull request by oeyh**
### Description
Adds some enhancements to to map_to_list processor:
- adds a `convert_field_to_list` option. If true, will convert fields to lists (`[key, value]`) instead of objects (`{key: value}`)
- allows source to be root when `source` is not specified
- adds `tags_on_failure` option to add tags when events failed to process
 
### Issues Resolved
Contributes to #3965 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java [hunk 1]
```diff
@@ -49,36 +52,75 @@ public Collection<Record<Event>> doExecute(final Collection<Record<Event>> recor
...
             }
         }
         return records;
     }
 
+    private Map<String, Object> getSourceMap(Event recordEvent) throws JsonProcessingException {
+        final Map<String, Object> sourceMap;
+        if (config.getSource() == null) {
+            // Source is root
+            sourceMap = OBJECT_MAPPER.treeToValue(recordEvent.getJsonNode(), Map.class);
```
Comment (asifsmohammed) [inline data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:93]: I have a high level question here, if you get the entire JSON node if source is null, I think it's not going to work. Correct me if I am wrong. 
For example, if the following is the event.
```
{
  "my-map": {
    "key1": "value1",
    "key2": "value2",
    "key3": "value3"
  }
}
```

The result will be
```
[
  ["my-map", {
    "key1": "value1",
    "key2": "value2",
    "key3": "value3"
  }]
]
```

Comment (kkondaka) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:93]: Good point. I think adding test cases and documentation for euch cases would help.

**Review by kkondaka — COMMENTED** [review]
Please document new config options in README.md. Also add examples for each option.

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java (hunk 1 above)
Comment (oeyh) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:93]: @asifsmohammed Good question, though it isn't necessarily related to root being the source. I think it's an issue with nested map in the source map. Currently we assume the map structure is flat. We can have a recursive option to go deeper into each field in the future if needed.

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java [hunk 2]
```diff
@@ -49,36 +52,75 @@ public Collection<Record<Event>> doExecute(final Collection<Record<Event>> recor
...
-                //TODO: add tagging on failure
+                recordEvent.getMetadata().addTags(config.getTagsOnFailure());
             }
         }
         return records;
     }
 
+    private Map<String, Object> getSourceMap(Event recordEvent) throws JsonProcessingException {
+        final Map<String, Object> sourceMap;
+        if (config.getSource() == null) {
```
Comment (dlvenable) [inline data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:91]: Is this a desirable feature?

If so, can we make it take in `/` instead of `null`? I think this could lead to confusion when users don't set the `source` field they are probably not getting what they want.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java (hunk 2 above)
Comment (oeyh) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:91]: Yes, we will need it for #3965. 

> If so, can we make it take in / instead of null? I think this could lead to confusion when users don't set the source field they are probably not getting what they want.

In [JsonPointer specification](https://datatracker.ietf.org/doc/html/draft-ietf-appsawg-json-pointer-03#section-5), it is actually the empty string `""` that refers to the whole document and `"/"` refers to the field with `""` as key. If we remove the restrictions in JacksonEvent that key cannot be empty, then we should already support referring to the root document with Json Pointer `""`. What do you think?

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java (hunk 1 above)
Comment (oeyh) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:93]: Added readme and more test cases.

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java (hunk 2 above)
Comment (dlvenable) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java:91]: @oeyh , That is a very useful finding.

I think we will need to make the syntax be explicit then:

```
source: ''
```

Comment (oeyh) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java]: Implemented this in recent commits.

**Review by dlvenable — DISMISSED** [review]

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java [hunk 3]
```diff
@@ -201,8 +200,13 @@ public <T> T get(final String key, final Class<T> clazz) {
...
     private JsonNode getNode(final String key) {
-        final JsonPointer jsonPointer = toJsonPointer(key);
-        return jsonNode.at(jsonPointer);
+        final String jsonPointerKey;
+        if (key.isEmpty() || key.startsWith("/")) {
+            jsonPointerKey = key;
+        } else {
+            jsonPointerKey = SEPARATOR + key;
+        }
+        return jsonNode.at(jsonPointerKey);
```
[the remark this answers is not in the document]
Comment (oeyh) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java]: Good point. Moved the changes to `toJsonPointer()` method. In the process, found some corner case issues with previous changes, addressed those and added more unit tests.

**Review by kkondaka — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
